src/main/diagnostics/applicabilityActionProvider.ts

```diff
@@ -71,45 +71,38 @@ export class ApplicabilityActionProvider extends AbstractFileActionProvider impl
-            for (const [region, severity] of topSeverityMap) {
-                // Add gutter icons of top severity for the region with the issue
-                this.addGutter(textEditor, SeverityUtils.getIcon(severity), region);
-            }
-            this._diagnosticCollection.set(document.uri, diagnostics);
+        const tree: CodeFileTreeNode | undefined = this._treesManager.issuesTreeDataProvider.getCodeIssueTree(document.uri.fsPath);
+        if (!tree) {
+            return;
         }
+        this._treesManager.logManager.logMessage("Creating applicable diagnostics for issue '" + tree.fullPath + "'", 'DEBUG');
+        this.addUnderlineToIssues(document, tree.issues);
+        this.addGutterToIssues(document, tree.issues);
     }
 
-    /**
-     * Generate diagnostics information for an applicable issue
-     * @param issue - the applicable issue to generate diagnostics for
-     * @param diagnostics - list of all the diagnostics of the file
-     * @param topSeverityMap - a map from regions in the file to their top severity
-     */
-    generateInformation(issue: IssueTreeNode, diagnostics: vscode.Diagnostic[], topSeverityMap: Map<vscode.Range, Severity>): void {
-        if (issue instanceof CodeIssueTreeNode) {
-            // Calculate top severity for region
-            let rangeSeverity: Severity | undefined = topSeverityMap.get(issue.regionWithIssue);
-            if (!rangeSeverity || rangeSeverity < issue.severity) {
-                topSeverityMap.set(issue.regionWithIssue, issue.severity);
-            }
-            // Add diagnostics
-            this._treesManager.logManager.logMessage("Creating applicable diagnostics for issue '" + issue.issueId + "'", 'DEBUG');
-            // Create diagnostics and gutter icon for the dependency
-            let created: vscode.Diagnostic[] = this.createDiagnostics(
-                issue.issueId,
-                'Severity: ' + SeverityUtils.getString(issue.severity),
-                vscode.DiagnosticSeverity.Warning,
-                issue.regionWithIssue
-            );
-            diagnostics.push(...created);
+    private addUnderlineToIssues(document: vscode.TextDocument, issues: CodeIssueTreeNode[]) {
+        this._diagnosticCollection.set(
+            document.uri,
+            issues.map(issue => super.issueToDiagnostic(issue))
+        );
+    }
+
+    private async addGutterToIssues(document: vscode.TextDocument, issues: CodeIssueTreeNode[]) {
+        let gutters: Map<vscode.Range, Severity> = this.issuesToGutters(issues);
+        const textEditor: vscode.TextEditor = await vscode.window.showTextDocument(document, vscode.ViewColumn.One);
+        for (const [region, severity] of gutters) {
+            this.addGutter(textEditor, SeverityUtils.getIcon(severity), region);
         }
     }
+
+    private issuesToGutters(issues: CodeIssueTreeNode[]): Map<vscode.Range, Severity> {
```

Comment:
maybe change the name to 'calculateTopSeverityForRegions' ? something that will better describe what the method does.

Reply:
Renamed to 'filterByTopSeverity'